### ISSUE TITLE
Can't Update Nor Delete AWS Toolkit for Visual Studio 2017

### ISSUE THREAD

**Issue by tankthomas** [aws/aws-sdk-net#783]
I'm using vs2017 and current tool version is 1.12.1.6,  running on window 10

See attached log:
[New Text Document.txt](https://github.com/aws/aws-sdk-net/files/1415191/New.Text.Document.txt)

**Comment by bloritsch** [comment]
I'm having the same issue.
[VSIXInstaller_b2a97691-0c0e-4704-89c9-ab2d86ddd170.log](https://github.com/aws/aws-sdk-net/files/1416577/VSIXInstaller_b2a97691-0c0e-4704-89c9-ab2d86ddd170.log)

**Comment by normj** [comment]
Can you look in the about box to get the exact version of Visual Studio? For example I'm currently running 15.3.4

**Comment by tankthomas** [comment]
I'm running 15.4.1. But the issue already exists on former version VS which is 15.3.x I don't remember exactly which version.

**Comment by bloritsch** [comment]
I'm also on 15.4.1.  Something happened after one of the updates where this plugin doesn't want to update, uninstall, etc.

**Comment by steveataws** [comment]
I just took the 15.4.1 update (having uninstalled the toolkit first) to try and repro errors on initial install, but didn't experience any problems. There was an oddity of the Tools and Extensions dialog listing two identically versioned copies of the toolkit. No-one else has reported this until now (and a quick test by normj@ showed only one copy listed when he searched - interested if anyone else sees this). The error in the log file attached at the head of this issue is something to do with our toolkit, based on the component id listed, but we're not sure what 'Catalog' it is referring to. In addition the only pre-requisite we note in our manifest (and which is the same for all packages) is on Microsoft.VisualStudio.Component.CoreEditor. Did any of you have any pre-RTM preview versions of the IDE installed? I'm wondering if the 'Catalog' being inspected isn't the one matching the instance the installer is trying to modify once it has determined the instance to upgrade.

I think we're going to have to try manually cleaning up your VS installation, which is tedious but might help Microsoft's installer unblock itself (and if it doesn't, maybe we'll get some more log data we can pass onto them). We've not made any changes to the toolkit specific to point versions of the IDE, and our package manifest is pretty simple, so we're unsure why we keep getting these random install problems since 15.3 shipped. We did hear back from Microsoft recently of changes to their installer to fix issues if the build tools had been installed, but I don't think that applies here.

To do a cleanup, try these steps:
1. Shutdown all running versions of Visual Studio.
2. Navigate to your VS install extensions folder (there may be multiple ones of these based on what skus you have installed). I have Professional, which is located by default at C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\Common7\IDE\Extensions. 
3. You'll see a bunch of random-named subfolders. For each subfolder, inspect it (or search for) files with name AWSToolkit*. If you see any, delete the subfolder containing the files. For example, on my install the toolkit was found at C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\Common7\IDE\Extensions\s1csxtxt.ydx. You need to check all subfolders (with the random names) as VS2017 has a deferred deletion scheme, so old versions can hang around 'deactivated'.
4. Go to the MSBuild folder (mine is at C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild - again, one per sku). Delete the folder named 'Amazon Web Services LLC' which holds the CloudFormation build targets.
5. Once you've cleaned the extensions folder and MSBuild folders, repeat for any other 2017 skus you may have installed (Enterprise, Community etc). 
6. Now navigate to your %USERPROFILE%/AppData/Local/Microsoft/VisualStudio folder. You'll see one or more subfolders prefixed '15.0_'. Go into each one and delete the ComponentModelCache folder (VS will rebuild it on restart).

Once all the steps above are done open a VS2017 command prompt in administrator mode and run the command 'devenv /setup'. Close the command prompt when done and then start VS2017 as you normally do. Verify the toolkit isn't shown in the About dialog (and as a menu item on the View menu). If all seems ok, head to the Tools and Extensions dialog and try a fresh install of the toolkit. Be sure to let us know how you get on.

**Comment by steveataws** [comment]
To be clear, you're not seeing 'AWS Explorer' under the View menu after the installer exits, nor any sign of the toolkit in the About dialog?

**Comment by tankthomas** [comment]
Yes. No sign of those things. Also, it doesn't exist in "extensions and updates". Repair visual studio can't solve the issue. 
Anyway, I uninstalled my vs and reinstalled it and everything works now.

**Comment by steveataws** [comment]
Ok, at least you're up and running which is the most important thing. Thanks for letting us know how the cleanup/reinstall went.

**Comment by steveataws** [comment]
I don't think there is much else to do here so resolving the issue. Feel free to reopen, or to open a new issue, if you encounter further issues with installation.

**Comment by viper25** [comment]
Visual Studio 2017 was crashing after I had created an AWS Lambda project and was deploying. I couldn't launch VS since then. Following the steps mentioned by @steveataws I was able to regain control of VS. 👍

**Comment by TylerDurden12** [comment]
I couldn't update or delete AWS Toolkit for Visual Studio (Tools > Extensions and Updates)

But the following step worked prety well for me :)
Programs and Features  > AWS Tools for Windows > Uninstall
https://docs.aws.amazon.com/toolkit-for-visual-studio/latest/user-guide/setup.html#uninstall

**Comment by joedean3** [comment]
It disappeared from the menu.   I uninstalled it, tried reinstalling it and it said it was already installed... Then I did a Visual Studio repair and it was back without reinstalling....  very strange.

**Comment by tman12345** [comment]
I was able to uninstall the tools by running the following.
1)	 Launch Visual Studio Installer
2)	Go to the VS you are using the n go to modify.
3)	Go to individual components  
a.	Then go all the way down to the bottom where it has Uncategorized section
b.	Uncheck AWS toolkit for Visual Studio 2017
c.	Click modify
It should now uninstall the tool so you can install again if you need to.

**Comment by artrodriguezcabrera** [comment]
I ran into this issue and the cleanest way to uninstall was following the steps @tman12345 documented.  Thank you!
